src/test_fixture/config.rs

```diff
@@ -0,0 +1,98 @@
+#![cfg(feature = "web-app")]
+
+use crate::config::{NetworkConfig, ServerConfig};
+use std::{fmt::Debug, net::TcpListener};
+
+pub const DEFAULT_TEST_PORTS: [u16; 3] = [3000, 3001, 3002];
+
+pub struct TestConfigBuilder {
+    ports: Option<[u16; 3]>,
+}
+
+impl TestConfigBuilder {
+    #[must_use]
+    pub fn with_default_test_ports() -> Self {
+        Self {
+            ports: Some(DEFAULT_TEST_PORTS),
+        }
+    }
+
+    #[must_use]
+    pub fn with_open_ports() -> Self {
+        Self { ports: None }
+    }
+
+    #[must_use]
+    pub fn with_specified_ports(ports: [u16; 3]) -> Self {
+        Self { ports: Some(ports) }
+    }
+
+    /// # Panics
+    /// If the configuration calls for automatic port assignment and binding sockets fails.
+    #[must_use]
+    pub fn build(self) -> Config {
+        let mut sockets = None;
+        let ports = self.ports.unwrap_or_else(|| {
+            let socks = [(); 3].map(|_| TcpListener::bind("127.0.0.1:0").unwrap());
+            let ports = socks
+                .iter()
+                .map(|sock| sock.local_addr().unwrap().port())
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap();
+            sockets = Some(socks);
```

Comment:
there is still a chance for port collision here, we drop the socket at the end and kernel is free to release it between the call to `build` and `listen`. it certainly does not happen often, but it may be nice to have a generic solution, similar to https://docs.rs/axum-server/latest/axum_server/struct.Handle.html